contracts/truefi2/AaveBaseRateOracle.sol

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.6.10;
+
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import {IAaveLendingPool} from "./interface/IAave.sol";
+
+contract AaveBaseRateOracle {
+    using SafeMath for uint256;
+
+    uint16 public constant BUFFER_SIZE = 365;
+    uint256 private constant BASIS_PRECISION = 10000;
+
+    // A cyclic buffer structure for storing running total (cumulative sum)
+    // values and their respective timestamps.
+    // latestIndex points to the previously inserted value.
+    struct RunningTotalsBuffer {
+        uint256[BUFFER_SIZE] runningTotals;
+        uint256[BUFFER_SIZE] timestamps;
+        uint16 currIndex;
+        uint256 lastValue;
+    }
+    RunningTotalsBuffer public totalsBuffer;
+
+    IAaveLendingPool public aavePool;
+
+    address asset;
+
+    // A fixed amount of time to wait
+    // to be able to update the totalsBuffer
+    uint256 public cooldownTime;
+
+    /**
+     * @dev Throws if cooldown is on when updating the totalsBuffer
+     */
+    modifier offCooldown() {
+        // get the last timestamp written into the buffer
+        uint256 lastWritten = totalsBuffer.timestamps[totalsBuffer.currIndex];
+        require(block.timestamp >= lastWritten.add(cooldownTime), "AaveBaseRateOracle: Buffer on cooldown");
+        _;
+    }
+
+    constructor(
+        IAaveLendingPool _aavePool,
+        uint256 _cooldownTime,
+        address _asset
+    ) public {
+        aavePool = _aavePool;
+        cooldownTime = _cooldownTime;
+        asset = _asset;
+
+        totalsBuffer.lastValue = getAaveVariableBorrowAPY(asset);
+        totalsBuffer.timestamps[0] = block.timestamp;
+    }
+
+    function bufferSize() public virtual pure returns (uint16) {
+        return BUFFER_SIZE;
+    }
+
+    /**
+     * @dev Helper function to get contents of the totalsBuffer
+     */
+    function getTotalsBuffer()
+        public
+        view
+        returns (
+            uint256[BUFFER_SIZE] memory,
+            uint256[BUFFER_SIZE] memory,
+            uint16
+        )
+    {
+        return (totalsBuffer.runningTotals, totalsBuffer.timestamps, totalsBuffer.currIndex);
+    }
+
+    function getAaveVariableBorrowAPY(address _asset) public view returns (uint256) {
```

Comment:
Since oracle works for one asset only, why do we pass it here? Cannot we just use the contract variable? Also, cannot we make it internal? (We can still make a test version where we override it and make it public. Also it would be really nice to have this function tested on mainnet fork.)